extensions/amp-google-assistant-assistjs/0.1/amp-google-assistant-voice-bar.js

```diff
@@ -62,10 +66,22 @@ export class AmpGoogleAssistantVoiceBar extends AMP.BaseElement {
       this.element.appendChild(iframe);
     });
 
+    let serviceHandlersMap = new Map();
+    serviceHandlersMap.set('RuntimeService.TriggerSendTextQuery',
+      () => {
+        this.frameService_.sendTextQuery();
+      });
+    serviceHandlersMap.set('RuntimeService.TriggerOpenMic',
+      () => {
+        this.frameService_.openMic();
+      });
+
     iframe.addEventListener('load', () => {
-      // TODO: create a channel to receive requests from underlying assist.js iframe.
-      this.frameService_.openMic();
-      this.frameService_.sendTextQuery();
+      this.channel_ = closure.createRespondingChannel(
```

Comment:
Sorry, I still need to push back on including the closure library. If it's absolutely necessary, I'll need to review the bundle that's being generated to make sure it's not mutating globals or installing unexpected polyfills.

What's actually needed here? It seems to just be setting up a [MessageChannel](https://developer.mozilla.org/en-US/docs/Web/API/MessageChannel) to pass to the iframe. If so, can we just manually create that and pass it to the iframe? That will save us the 15kb of code, and save me from having to review it 😉.

Reply:
Sorry for the inconvenience and yes I totally understand the compiled Closure code is definitely not readable at all... Is there any other way we could include the Closure library? As I mentioned in the PR description above, the Closure library in npm is not importable/usable since it is a node.js library. And this won't be the first PR to add 3P libraries. We have something like PortOperator(also from Closure library) and Protocol Buffer library to be added.  

There are several advantages of using Closure library: 
1. handles the repetitive details of registering and dispatching to services(provided by AbstractChannel extended by PortChannel)
2. provides a robust way to create channels and send message ports between parent document and iframes(provided by PortChannel)
3. provides a more clear(more API-like) way to register services and get back responses(provided by RespondingChannel wrapping a MessageChannel)
4. all existing assist.js communication between runtime(main document) and iframes are based on PortChannel and RespondingChannel
5. PortOperator(not included/mentioned in this PR but also in Closure) is needed for direct inter-iframe communication. It coordinates and assigns port to children iframes so that children iframes could talk to each other directly without going through runtime(main document) first.

@ianba feel free to add more details here if I am missing/misunderstanding anything here.

Reply:
You can also see more details about the implementation of Closure messaing package here: https://github.com/google/closure-library/tree/master/closure/goog/messaging.

Reply:
With the changes I suggested in the compiler (and the bundle recompiled), this doesn't install any polyfills. However, there's an `eval` call (which won't run in our CSP), a code transpiler (!!!), a sync `XMLHttpRequest` (which is being banned by the web platform), a `<script>` with custom JS endpoints (which won't load in our CSP).

I seriously doubt the code will actually work when served from the AMP Cache, and it's way overkill for a simple message passing. If we need to use a `MessageChannel`, that's fine. We need to clearly define what messages expect to receive on it, and what messages we need to send. Then we hand write the code that does only that. I expect we'll go from 15kb of compressed code to 500b.

Reply:
A couple of thoughts:
1) Let's keep looking at the config. We definitely don't use eval, XHRs, or download <script>s with this. Certainly we can cut some stuff. Probably disabling the debug flag will help... some of the script loading stuff seemed associated with that. Also we don't need logclient/logserver from the messaging library and probably some other classes too.

2) IIUC this code is only loaded by pages that include our extension. So while not ideal, it's not going to break all of AMP. Currently this is an experimental project. Assuming this extra 10kb doesn't throw performance off a cliff, I'd rather us get the thing working, prove its value, and then optimize. Yes we could hand write a messaging library just for AMP, but it's going to take time to implement and require extra maintenance to keep up to date w/ the messaging in the non-AMP version. Does it seem reasonable to make that tradeoff and revisit this for optimization later?

Reply:
@jridgewell what do you think? Let's revisit the optimization after we want to graduate this fomr an experimental project?

For the concerning/illegal usage of code you mentioned above, they are all from the base.js in Closure(base.js is also a big reason why the compiled code grows to this big...). We are unable to get rid of it since it's the entry point of the entire library. However, I read though the code make sure some of them are not called at all. I also redefined some flags in the Closure compiler to make sure those code won't be called:
* base.js
  * eval
     * loadModuleFromSource_: redefine goog.LOAD_MODULE_USING_EVAL as false so the function won't be called in goog.loadModule
  * Transpiler used in all types of dependencies: goog.TRANSPILE set to 'never'
     * TransformedDependency, TranspiledDependencies, and other dependencies specified in base.js are only for DebugLoader. Dependencies would be added via createDependency in DependencyFactory under DebugLoader. However, DebugLoader won't be enabled at all under compiled mode according to goog.DEPENDENCIES_ENABLED.
  * Sync XMLHttpRequest
     * loadFileSync called in TransformedDependency: TransformedDependency is only used for goog.modules and Closure library doesn't have any of it. 
     * loadFileSync called in goog.transpile_: same reason above, any transpile logic won't be called because DebugLoader is disabled.
  * script tag with custom JS endpoints
     * writeScriptTag_: this is used by getLoader and explorer.js which is never included in the compiled code
     * Dependency.load: same reason above, dependency related logic in DebugLoader won't be called at all since DebugLoader is disabled.
     * TransformedDependency.load: same as above

Reply:
Without enough clarity around the usage of `eval`, transpilation, sync xmlhttprequest, and custom endpoints I believe we should ask for a security review related to usage of this code on an AMP document.

@ampproject/wg-security-privacy **Can you please take a look at this request and advise on this matter?**

Additionally, if we can prove that none of these codepaths actually execute on a real document I'd ask for two things:
1. Why must this code be transmitted to clients if it never executes?
2. Is it possible that a minor update to logic will cause it to unintentionally start executing? 

If these two questions can be clearly answered with the blessing of the @ampproject/wg-security-privacy approval, then shipping this change locally only to this extension for a short period of time makes sense. This time period should be long enough to learn about the value of the extension, but not longer than necessary to make a call if development should continue.

If the signal is clear that development should continue, or usage is expected ongoing then we'd ask for the code present on the main frame of an AMP document follow the patterns expected elsewhere in AMP and according to the design philosophies it follows. This means the `Closure` library must either strip these references or an implementation that supports the needs of the extension and nothing further should be developed.

Reply:
@molnarg FYI

Reply:
@ychsieh I looked at the code we serve for the non-AMP assist.js runtime and it doesn't have any of this stuff (`eval`, etc) even though we are using this library. Can you please try using the `RECOMMEND_FLAGS` we use to compile it here as well? Hopefully should get rid of the unused code.

Reply:
I tried the RECOMMEND_FLAGS and it does remove all the strange stuff like eval, transpilation, xmlhttprequest, etc. But seems like those flags are not accepted using the public Closure Compiler. There might be some usage differences between the internal and external compiler, e.g. Google-only code in the internal one. What I did with the latest commit is to replace the closure-bundle.js with the locally compiled code using internal Google build tools. 

I think I am going to compare the internal JSCompiler and the external Closure compiler to try to find the differences that cause/remove the unnecessary code. Hopefully we can extend the public compiler and put the extended version here to compile without the unnecessary code. What do you guys think?

Reply:
So currently I asked a question on YAQS about the best way to compile the code with the RECOMMENDED_FLAGS in a open-source world, either update the public one or use the private one directly. 

In the mean time, is it okay to just check in the current updated compiled code since those unsafe stuff are removed and the size is also reduced?

Reply:
Awesome work! This LGTM, minus:

I scanned the code again, and the only problematic part of the code is:

```js
var X = ["createRC"],
  Y = m;
X[0] in Y || "undefined" == typeof Y.execScript || Y.execScript("var " + X[0]);
for (var Z; X.length && (Z = X.shift()); )
  X.length || void 0 === cc
    ? Y[Z] && Y[Z] !== Object.prototype[Z]
      ? (Y = Y[Z])
      : (Y = Y[Z] = {})
    : (Y[Z] = cc);
```
(It's at the end). If I'm reading this right, it's going to install `Y[Z] = cc`, where `Y === window`, `Z === "createRC"`, and `cc` is the `createRc` function. This conflicts with AMP globals requirement, which requires that all installed globals follow the pattern `__AMP_[A-Z_]+` (there's a script vulnerability with globals since any DOM node can override it with `name=createRc` or `id=createRC` and we don't control what's in the HTML).

I believe this is coming from the `goog.exportSymbol('createRC', createRC);`. Can this be removed, or can we move the `createRC` call to the AMP side instead of the closure output?

Reply:
Sounds like we're close to a resolution here. Do we still need a security review?

Reply:
@jridgewell Unfortunately that function has to be exported to work. I just renamed it to __AMP_createRC. The name doesn't matter since the final endpoint is createRespondingChannel.

Thanks a lot for pinpointing the issue and replied on the YAQS! That really saves our time to go through the differences between the public and internal Closure compiler. 

@kristoferbaxter yes we don't need a security review anymore since the irrelavant stuff like eval and transpiler are removed in the newest compiled code.

Reply:
@kristoferbaxter: Agree, security review isn't necessary anymore.

Reply:
Awesome, thanks for tracking this down all.